File: packages/depz-sensor-sdk-java/src/main/java/ai/depz/sensor/sensors/bno086/Sh2.java
package ai.depz.sensor.sensors.bno086;

/**
 * SH-2 control-channel encoders (contracts/05_SENSOR_BNO086.md §6). Pure
 * codecs; mirror of the TS/Python {@code sh2} references. Report/response
 * decoders that require a live hub are out of the verifiable-vector scope.
 */
public final class Sh2 {
    private Sh2() {}

    // Control report IDs.
    public static final int SET_FEATURE_COMMAND = 0xfd;
    public static final int GET_FEATURE_REQUEST = 0xfe;
    public static final int GET_FEATURE_RESPONSE = 0xfc;
    public static final int PRODUCT_ID_REQUEST = 0xf9;
    public static final int PRODUCT_ID_RESPONSE = 0xf8;
    public static final int COMMAND_REQUEST = 0xf2;
    public static final int COMMAND_RESPONSE = 0xf1;
    public static final int FRS_READ_REQUEST = 0xf4;
    public static final int FRS_READ_RESPONSE = 0xf3;
    public static final int FRS_WRITE_REQUEST = 0xf7;
    public static final int FRS_WRITE_DATA = 0xf6;
    public static final int FRS_WRITE_RESPONSE = 0xf5;

    /** Set Feature Command (0xFD), 17 bytes. intervalUs = 0 disables. */
    public static byte[] buildSetFeature(
            int sensorId, long intervalUs, long batchUs, int sensitivity, int flags, long cfgWord) {
        byte[] out = new byte[17];
        out[0] = (byte) SET_FEATURE_COMMAND;
        out[1] = (byte) sensorId;
        out[2] = (byte) flags;
        putU16le(out, 3, sensitivity);
        putU32le(out, 5, intervalUs);
        putU32le(out, 9, batchUs);
        putU32le(out, 13, cfgWord);
        return out;
    }

    /** Get Feature Request (0xFE), 2 bytes. */
    public static byte[] buildGetFeatureRequest(int sensorId) {
        return new byte[] {(byte) GET_FEATURE_REQUEST, (byte) sensorId};
    }

    /** Product ID Request (0xF9), 2 bytes. */
    public static byte[] buildProductIdRequest() {
        return new byte[] {(byte) PRODUCT_ID_REQUEST, 0x00};
    }

    /** Command Request (0xF2), 12 bytes: id, seq, command, P0..P8. */
    public static byte[] buildCommandRequest(int seq, int command, byte[] params) {
        if (params.length > 9) {
            throw new IllegalArgumentException("command request carries at most 9 parameter bytes");
        }
        byte[] out = new byte[12];
        out[0] = (byte) COMMAND_REQUEST;
        out[1] = (byte) seq;
        out[2] = (byte) command;
        System.arraycopy(params, 0, out, 3, params.length);
        return out;
    }

    /** FRS Read Request (0xF4), 8 bytes. blockWords = 0 reads the record. */
    public static byte[] buildFrsReadRequest(int frsType, int offsetWords, int blockWords) {
        byte[] out = new byte[8];
        out[0] = (byte) FRS_READ_REQUEST;
        out[1] = 0;
        putU16le(out, 2, offsetWords);
        putU16le(out, 4, frsType);
        putU16le(out, 6, blockWords);
        return out;
    }

    /** FRS Write Request (0xF7), 6 bytes. lengthWords = 0 erases the record. */
    public static byte[] buildFrsWriteRequest(int frsType, int lengthWords) {
        byte[] out = new byte[6];
        out[0] = (byte) FRS_WRITE_REQUEST;
        out[1] = 0;
        putU16le(out, 2, lengthWords);
        putU16le(out, 4, frsType);
        return out;
    }

    /** FRS Write Data (0xF6), 12 bytes; 1 or 2 words per packet. */
    public static byte[] buildFrsWriteData(int offsetWords, long[] words) {
        if (words.length < 1 || words.length > 2) {
            throw new IllegalArgumentException("FRS write data carries 1 or 2 words");
        }
        byte[] out = new byte[12];
        out[0] = (byte) FRS_WRITE_DATA;
        out[1] = 0;
        putU16le(out, 2, offsetWords);
        putU32le(out, 4, words[0]);
        putU32le(out, 8, words.length > 1 ? words[1] : 0);
        return out;
    }

    private static void putU16le(byte[] b, int off, int v) {
        b[off] = (byte) v;
        b[off + 1] = (byte) (v >>> 8);
    }

    private static void putU32le(byte[] b, int off, long v) {
        b[off] = (byte) v;
        b[off + 1] = (byte) (v >>> 8);
        b[off + 2] = (byte) (v >>> 16);
        b[off + 3] = (byte) (v >>> 24);
    }
}

File: packages/depz-sensor-sdk-rust/src/bno086/sh2.rs
//! SH-2 control-channel request builders (contract 05 §6).
//!
//! Pure encoders — no I/O, no timing. The SHTP framing/seq is [`super::shtp`]'s
//! concern. All multi-byte fields little-endian.
//!
//! Mirrors the TS/Python `sh2` builders.

/// Report IDs on SHTP channel 2 (control).
pub const REPORT_COMMAND_RESPONSE: u8 = 0xf1;
pub const REPORT_COMMAND_REQUEST: u8 = 0xf2;
pub const REPORT_FRS_READ_RESPONSE: u8 = 0xf3;
pub const REPORT_FRS_READ_REQUEST: u8 = 0xf4;
pub const REPORT_FRS_WRITE_RESPONSE: u8 = 0xf5;
pub const REPORT_FRS_WRITE_DATA: u8 = 0xf6;
pub const REPORT_FRS_WRITE_REQUEST: u8 = 0xf7;
pub const REPORT_PRODUCT_ID_RESPONSE: u8 = 0xf8;
pub const REPORT_PRODUCT_ID_REQUEST: u8 = 0xf9;
pub const REPORT_GET_FEATURE_RESPONSE: u8 = 0xfc;
pub const REPORT_SET_FEATURE_COMMAND: u8 = 0xfd;
pub const REPORT_GET_FEATURE_REQUEST: u8 = 0xfe;

/// SH-2 command builder error.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Sh2Error {
    /// Command request carries at most 9 parameter bytes.
    TooManyParams(usize),
    /// FRS write data carries 1 or 2 words.
    BadWriteWordCount(usize),
}

impl std::fmt::Display for Sh2Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Sh2Error::TooManyParams(n) => {
                write!(f, "command request carries at most 9 params, got {}", n)
            }
            Sh2Error::BadWriteWordCount(n) => write!(f, "FRS write data carries 1 or 2 words, got {}", n),
        }
    }
}

impl std::error::Error for Sh2Error {}

/// Set Feature Command (0xFD), 17 bytes. `interval_us` = 0 disables the sensor.
pub fn build_set_feature(
    sensor_id: u8,
    interval_us: u32,
    batch_us: u32,
    sensitivity: u16,
    flags: u8,
    cfg_word: u32,
) -> [u8; 17] {
    let mut out = [0u8; 17];
    out[0] = REPORT_SET_FEATURE_COMMAND;
    out[1] = sensor_id;
    out[2] = flags;
    out[3..5].copy_from_slice(&sensitivity.to_le_bytes());
    out[5..9].copy_from_slice(&interval_us.to_le_bytes());
    out[9..13].copy_from_slice(&batch_us.to_le_bytes());
    out[13..17].copy_from_slice(&cfg_word.to_le_bytes());
    out
}

/// Get Feature Request (0xFE), 2 bytes.
pub fn build_get_feature_request(sensor_id: u8) -> [u8; 2] {
    [REPORT_GET_FEATURE_REQUEST, sensor_id]
}

/// Product ID Request (0xF9), 2 bytes.
pub fn build_product_id_request() -> [u8; 2] {
    [REPORT_PRODUCT_ID_REQUEST, 0x00]
}

/// Command Request (0xF2), 12 bytes: id, seq, command, P0..P8.
pub fn build_command_request(seq: u8, command: u8, params: &[u8]) -> Result<[u8; 12], Sh2Error> {
    if params.len() > 9 {
        return Err(Sh2Error::TooManyParams(params.len()));
    }
    let mut out = [0u8; 12];
    out[0] = REPORT_COMMAND_REQUEST;
    out[1] = seq;
    out[2] = command;
    out[3..3 + params.len()].copy_from_slice(params);
    Ok(out)
}

/// FRS Read Request (0xF4), 8 bytes. `block_words` = 0 reads the whole record.
pub fn build_frs_read_request(frs_type: u16, offset_words: u16, block_words: u16) -> [u8; 8] {
    let mut out = [0u8; 8];
    out[0] = REPORT_FRS_READ_REQUEST;
    out[1] = 0;
    out[2..4].copy_from_slice(&offset_words.to_le_bytes());
    out[4..6].copy_from_slice(&frs_type.to_le_bytes());
    out[6..8].copy_from_slice(&block_words.to_le_bytes());
    out
}

/// FRS Write Request (0xF7), 6 bytes. `length_words` = 0 erases the record.
pub fn build_frs_write_request(frs_type: u16, length_words: u16) -> [u8; 6] {
    let mut out = [0u8; 6];
    out[0] = REPORT_FRS_WRITE_REQUEST;
    out[1] = 0;
    out[2..4].copy_from_slice(&length_words.to_le_bytes());
    out[4..6].copy_from_slice(&frs_type.to_le_bytes());
    out
}

/// FRS Write Data (0xF6), 12 bytes; 1 or 2 words per packet.
pub fn build_frs_write_data(offset_words: u16, words: &[u32]) -> Result<[u8; 12], Sh2Error> {
    if words.is_empty() || words.len() > 2 {
        return Err(Sh2Error::BadWriteWordCount(words.len()));
    }
    let mut out = [0u8; 12];
    out[0] = REPORT_FRS_WRITE_DATA;
    out[1] = 0;
    out[2..4].copy_from_slice(&offset_words.to_le_bytes());
    out[4..8].copy_from_slice(&words[0].to_le_bytes());
    out[8..12].copy_from_slice(&words.get(1).copied().unwrap_or(0).to_le_bytes());
    Ok(out)
}
